Add v2 REST API specification

iain-anderson opened the pull request:
Signed-off-by: Iain Anderson <iain@iotechsys.com>

## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] The commit message follows our guidelines: https://wiki.edgexfoundry.org/display/FA/Committing+Code+Guidelines#CommittingCodeGuidelines-CommitMessages
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?
<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ x] Documentation content changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior and link to a relevant issue. -->

Issue Number:


## What is the new behavior?


## Does this PR introduce a breaking change?
<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->

- [ ] Yes
- [ ] No


## Are there any specific instructions or things that should be known prior to reviewing?

## Other information

iain-anderson commented:
Note in particular:

- `device/all` is removed
- `callback` is replaced by:

- - `callback/device` (post/put)
- - `callback/device/id` (delete)
- - `callback/profile` (post/put)
- - `callback/profile/id` (delete)
- - `callback/watcher` (post/put)
- - `callback/watcher/id` (delete)

jpwhitemn commented:
LGTM Iain.  It looks like it shares much with the other OpenAPI docs including and most importantly command service.  One question, in this API spec, CommandAction items ref ActionResponse, whereas in command service API spec it ref's CommandResponse.  Was there a reason for this difference?

iain-anderson commented:
Hi Jim, it's because the core-metadata and core-command API specs name them differently :)
This inconsistency has gone away now that edgex-go #2447 has landed

iain-anderson commented:
Updated in line with https://github.com/edgexfoundry/edgex-go/pull/2447

lenny-intel commented on api/oas3.0/v2/device-sdk.yaml:
```diff
@@ -0,0 +1,1269 @@
+  /discovery:
+    post:
+      description: Run the discovery request for a Device Service.
+      responses:
```
RequestBody? i.e. for consistency we should have a request DTO for every API? If not, then document that not request body required.

iain-anderson replied:
Noted no request body

lenny-intel commented on api/oas3.0/v2/device-sdk.yaml:
```diff
@@ -0,0 +1,1269 @@
+      - $ref: '#/components/schemas/BaseResponse'
+      type: object
+      properties:
+        version:
```
For V1 the SDKs  are returning SDK version and Service version. Need to do the same for V2?

iain-anderson replied:
Yes, amended

iain-anderson commented on api/oas3.0/v2/device-sdk.yaml:
```diff
@@ -0,0 +1,1269 @@
+components:
+  schemas:
+    AutoEvent:
+      type: object
```
done

iain-anderson commented on api/oas3.0/v2/device-sdk.yaml:
```diff
@@ -0,0 +1,1269 @@
+        - type: object
+          properties:
+            binaryValue:
+              description: "If the value of the reading is binary, it will be found in this property as a byte array"
```
done

iain-anderson commented on api/oas3.0/v2/device-sdk.yaml:
```diff
@@ -0,0 +1,1269 @@
+          properties:
+            binaryValue:
+              description: "If the value of the reading is binary, it will be found in this property as a byte array"
+              type: string
```
Changed to format: binary

iain-anderson commented on api/oas3.0/v2/device-sdk.yaml:
```diff
@@ -0,0 +1,1269 @@
+        attributes:
+          type: object
+          additionalProperties:
+            type: string
```
I don't think so. These are fields with arbitrary names.

lenny-intel replied:
`additionalProperties` indicates it is a list of some type. CSV?

iain-anderson replied:
I think it works as a dictionary as highlighted here: https://swagger.io/docs/specification/data-models/dictionaries/

iain-anderson commented on api/oas3.0/v2/device-sdk.yaml:
```diff
@@ -0,0 +1,1269 @@
+      properties:
+        config:
+          description: "A string-ified representation of the service's configuration. For purposes of this specification, a string has been used since configuration structure differs from service to service."
+          type: string
```
done

iain-anderson commented on api/oas3.0/v2/device-sdk.yaml:
```diff
@@ -0,0 +1,1269 @@
+      description: "Notification that a new device profile has been created."
+      type: object
+      properties:
+        device:
```
Yes. copy-paste.

iain-anderson commented on api/oas3.0/v2/device-sdk.yaml:
```diff
@@ -0,0 +1,1269 @@
+      description: "Notification that a new provision watcher associated with this device service has been created."
+      type: object
+      properties:
+        device:
```
fixed

iain-anderson commented on api/oas3.0/v2/device-sdk.yaml:
```diff
@@ -0,0 +1,1269 @@
+    ProtocolProperties:
+      type: object
+      additionalProperties:
+        type: string
```
As above, don't think so

iain-anderson commented on api/oas3.0/v2/device-sdk.yaml:
```diff
@@ -0,0 +1,1269 @@
+          type: integer
+        identifiers:
+          type: object
+          additionalProperties:
```
As above, don't think so

iain-anderson commented on api/oas3.0/v2/device-sdk.yaml:
```diff
@@ -0,0 +1,1269 @@
+      description: "Notification that an existing device profile has been updated."
+      type: object
+      properties:
+        device:
```
done

iain-anderson commented on api/oas3.0/v2/device-sdk.yaml:
```diff
@@ -0,0 +1,1269 @@
+      description: "Notification that an existing provision watcher associated with this device service has been updated."
+      type: object
+      properties:
+        device:
```
done

iain-anderson commented on api/oas3.0/v2/device-sdk.yaml:
```diff
@@ -0,0 +1,1269 @@
+              schema:
+                $ref: '#/components/schemas/ConfigResponse'
+        '207':
+          description: "Indicates a multi-part response supportive of accepting multiple requests at once. The 'statusCode' property of each response in the returned array will indicate success or failure."
```
Don't know. Spec is taken from edgex-go. @tsconn23 ?

lenny-intel replied:
@tsconn23, Does this need to have a multi-part response?

tsconn23 replied:
Multi-part is recommended to establish a common pattern for all request/responses. In practice the endpoint would return an array with only one element, but it's better in my view to do that than have one (or a handful) endpoint with a different behavior.

lenny-intel replied:
Got it. Thanks!

iain-anderson commented:
Added descriptions. Added some "requires" specifications but many of these are context-dependent (eg the required fields for posting a new object are different to an update), am guessing this is why they aren't specified in edgex-go?

iain-anderson commented:
> Should it be our intent to have description for every object and every property?
The definitions in question are imported from edgex-go/api/openapi/v2 so I don't believe it's currently the intention. Perhaps it should be though.

iain-anderson commented:
Added additional description fields
